src/java.base/share/classes/jdk/internal/module/ModuleHashes.java

```diff
@@ -162,16 +166,28 @@ static byte[] computeHash(Supplier<ModuleReader> supplier, String algorithm) {
      * @throws UncheckedIOException if an I/O error occurs
      */
     static ModuleHashes generate(Set<ModuleReference> mrefs, String algorithm) {
-        Map<String, byte[]> nameToHash = new TreeMap<>();
-        for (ModuleReference mref : mrefs) {
-            try (ModuleReader reader = mref.open()) {
-                byte[] hash = computeHash(reader, algorithm);
-                nameToHash.put(mref.descriptor().name(), hash);
-            } catch (IOException ioe) {
-                throw new UncheckedIOException(ioe);
+        Map<String, byte[]> nameToHash = new ConcurrentHashMap<>();
+        mrefs.stream().parallel().forEach(mref -> {
```

Comment:
[AFAIK](https://stackoverflow.com/questions/28985704/parallel-stream-from-a-hashset-doesnt-run-in-parallel) streams taken from HashSet might have weak parallelism, so maybe it's worth wrapping `mrefs` into an ArrayList?